### PR TITLE
Smooth out testing behaviors

### PR DESCRIPTION
**Pull request by portante**
Filed as 2 separate commits:

 * Agent/server sources builds run when required
   * The agent and server sources builds now only by default, or when explicitly requested.  If a separate sub-test is requested, they are no longer gratuitously run.
 * Quiet the Jenkins podman cleanup step
   * We create a script to ensure the cleanup step exits successfully.

### CODE REVIEW

**Comment by portante** [comment]
Moving to `Draft` mode until PR #3048 lands.

**Review by npalaska — DISMISSED** [review]

**Review by dbutenhof — APPROVED** [review]

**Review by ndokos — APPROVED** [review]

File: exec-unittests
Comment (webbnh) [inline exec-unittests:185]: This file appears to use spaces for indentation, but line 185 (et al.) are using hard-tabs.  We should be consistent.

File: exec-unittests
Comment (webbnh) [inline exec-unittests:207]: Do we still want to do this, now that the CI is building the RPMs for the Agent and the Server, whose build process relies on the source-build?

File: jenkins/Pipeline.gy
Comment (webbnh) [inline jenkins/Pipeline.gy:70]: I'm not a fan of having lots of little scriptlets floating around, particularly ones which have specialized uses.

If we think that `podman-cleanup` is of general use, then perhaps we should put it in `.../utils`.  Otherwise, perhaps something like this would be better:
```
        always {
            // Remove the pod which we just created and ran; remove any
            // dangling containers; and then remove any dangling images.
            sh 'podman pod rm -f ${PB_POD_NAME} || true'
            sh 'podman container prune -f || true'
            sh 'podman image prune -f || true'
        }
```

**Review by webbnh — APPROVED** [review]
Looks OK, but I suspect that we should fix the indentation.  (I wonder if there is a tool like Black or iSort to deal with that sort of thing....)

Also, I have a question and a...preference.

File: exec-unittests
Comment (portante) [reply exec-unittests]: I like the immediate feedback that the source build breakage happens early in the pipeline and more accessible to the developer who doesn't have to run much to get the feedback.

If folks would rather this in one place, I'd be happy to rip it out.

**Review by dbutenhof — DISMISSED** [review]

**Review by webbnh — DISMISSED** [review]
Looks good...once we resolve the previous conversations.

File: exec-unittests
Comment (webbnh) [inline exec-unittests:184]: I _think_ the code at line 189 which checks for a zero count is now pointless, because `count` will only be zero if `_para_jobs_file` is empty, in which case we won't enter the block.

If we don't enter the block here, then we know that `count` would be zero, so we don't actually need to calculate it (so we can remove the `let` statement and the increment at lines 177 and 181, respectively), and we can move the `if` at line 189 to an `elif` to this `if`.  (Of course, we should remove the `count` test from its condition, and we presumably need to add a check for not `"srcbld"`.)

File: exec-unittests
Comment (webbnh) [reply exec-unittests]: No -- I can see the value of this -- we probably shouldn't rip it out entirely.  However, I'm not convinced that the immediacy is worth imposing the cost in all cases.

I noticed in my CI work that this step takes a non-trivial amount of time.  So, if (for example) we are going to proceed on to running the RPM build anyway, it's not helpful to perform this step here.

By the way, I noticed a couple of things.  First, we're performing this step even if the preceding step failed.  In the name of immediacy, it seems like we should be proceeding to report that failure instead of continuing with this step.  Second, I _think_ that the very next thing the CI does after this step is to build the RPMs, so doing this here (in that case) doesn't actually speed the reporting of the failure, it just delays the completion of the build.

So, I have two suggestions:  first, remove the defaulting -- if `srcbld` is requested, then run the step; second, add to the conditional so that this step is skipped if `rc` is non-zero.

**Review by webbnh — COMMENTED** [review]
Unfortunately, in looking at the `"srcbld"` stuff I found some now-dead code which we should probably address

File: exec-unittests
Comment (portante) [reply exec-unittests:192]: Fixed.

This pointed out other bad behaviors, though.  I made a number of updates so that we get the expected driver output, and corrected the list of unit tests sub-systems enumerated.

File: exec-unittests
Comment (portante) [reply exec-unittests]: I routine type `tox -e py39 -- agent` to verify all the agent code works.  I want to be sure that I don't break the source tree builds each time I do that.  If we remove `srcbld` as the default, I'll have to wait for the CI job failure to catch a problem.

I'd rather have a simple `cd agent/rpm; make rpm` verify it is all working, but that is out of the scope of this PR.

So as I think about this, I'd rather keep this as is and not remove the default.

**Review by portante — COMMENTED** [review]
I think this is good now.

**Review by npalaska — DISMISSED** [review]

File: exec-unittests
Comment (dbutenhof) [inline exec-unittests:136]: That's incorrect. `lib/pbench/test/unit/client` exists, with two test modules and a handful of test cases for create and connect:

```
(.venv39) dbutenho 07:40 audit:audit tox client
[...]
Running client python3-based unit tests via pytest

===================================================================== test session starts ======================================================================
platform linux -- Python 3.9.7, pytest-7.1.3, pluggy-1.0.0
cachedir: /var/tmp/dbutenho/tox/py39/.pytest_cache
rootdir: /home/dbutenho/github/audit, configfile: pytest.ini
plugins: requests-mock-1.10.0, mock-3.10.0, helpers-namespace-2019.1.8, freezegun-0.4.2, forked-1.4.0, cov-4.0.0, xdist-2.5.0
gw0 [24] / gw1 [24] / gw2 [24] / gw3 [24]
........................
```

**Review by dbutenhof — CHANGES_REQUESTED** [review]

**Review by riya-17 — APPROVED** [review]

File: exec-unittests
Comment (portante) [reply exec-unittests:136]: Ugh, I totally messed that up.  Fixing...

Comment (portante) [reply exec-unittests:136]: Restored.

But ... this is weird.

The client sub-system is really part of server.  The majors are the top-level directories, agent, server, and dashboard.

What are the "clients"?

Comment (dbutenhof) [reply exec-unittests:136]: I didn't spend a lot of time thinking about packaging at the time; I wanted to get a start on the basic framework for functional testing.

It's associated with the server, but the idea is that we want this code *outside* the server, not *inside* the server. It's a pseudo-separate package, and it has its own unit tests, but currently we're only building it as part of the server package. Yeah, that's really wrong, and it probably needs its own `setup.py` and packaging. Maybe we can think about that when we move to poetry.

File: exec-unittests
Comment (webbnh) [inline exec-unittests:148]: I can't figure it out:  when will `_pytest_majors` ever be empty at line 147?  (I'm inclined to suggest that you restore the change to line 130 and omit lines 147-148.)

File: exec-unittests
Comment (webbnh) [inline exec-unittests:182]: The default value is cool an' all, but it seems unnecessary:  if `subtst` is undefined or blank, it won't be equal to `python` anyway, so replacing it with `legacy` doesn't accomplish much (other than making me ponder why we're doing it 😉).  So, I suggest comparing `"${subtst}"` "as is" to `"python"`.

Ditto line 210.

(And, as long as we're giggling over this stuff, the quotation marks around `"python"` don't accomplish anything either -- we've just been programming in languages like C and Python for so long that they _seem_ like they are required.  😆)

File: exec-unittests
Comment (webbnh) [inline exec-unittests:199]: Don't like `elif`, eh?

**Review by webbnh — APPROVED** [review]
I'm approving, in case my concerns are wide of the mark, but I've noted a couple of things below that I'm confused/concerned about (and there are a couple of other suggestions, as well, in case you have reason to make more changes).

**Review by dbutenhof — DISMISSED** [review]

File: exec-unittests
Comment (portante) [reply exec-unittests:137]: Sweet, I was hoping for some good `bash` fu.

File: exec-unittests
Comment (portante) [reply exec-unittests:182]: I guessed this would come up.  =)

The reason I used the defaulting was to help me understand the flow through this.  We are being a bit tricky by allowing the `subtst` variable to remain empty to indicate a "do it all" behavior, but at times we need to default the value to make it easier to know what to do in a given case.

So I kept the defaulting in place for the expected context for the operation being considered.

File: exec-unittests
Comment (portante) [reply exec-unittests:137]: Dropping this entirely since I was way off on what the "major" `client` is.

Comment (webbnh) [reply exec-unittests:137]: > Sweet, I was hoping for some good `bash` fu.

I'm glad you liked it...with `bash`, it's often hard to tell what is "sweet" and what is unpleasant obfuscation.  🥴

File: exec-unittests
Comment (webbnh) [reply exec-unittests:182]: I get the part about having "empty" mean "all", but I don't see how applying a default of `legacy` here helps provide clarity.  (In fact, I'm inclined to argue that it does the opposite.)

I think what we're trying to express in this code is that, if `${subtst}` is, specifically, `python` and nothing else, then we should skip to the next iteration.  Saying first that if `${subtst}` is empty/all then treat it as though it were `legacy` (so that it won't match `python`) just makes the reader wonder, since empty already isn't `python`.

If you really want to apply a default here, how about using `all`, which would be semantically consistent with the intended use of empty, and that would fail to match `python` just as effectively as `legacy` would....

**Review by portante — COMMENTED** [review]
Applied all known review feedback.

File: exec-unittests
Comment (webbnh) [inline exec-unittests:134]: Nice!

File: exec-unittests
Comment (webbnh) [inline exec-unittests:175]: With this change, there is now no common code in the body of the loop -- it's segregated by `agent`/`client`/`server` -- in which case there's not much point in having the loop.  If you're amenable, instead we can just say "if `agent` is in `$_major`, then do the agent stuff", then "if `server` is in `$_major`, then do the server stuff", and so on.  (If you're interested and don't know how to check for a value in a list, I'll see what I can come up with....)

File: exec-unittests
Comment (webbnh) [inline exec-unittests:169]: This seems like a weird place to define this.  I would think that this belongs at the top of the file, as a "configuration definition".  Failing that, I would place it near where it is used.

File: exec-unittests
Comment (webbnh) [inline exec-unittests:183]: I think it's a bad idea to have two different variables whose names differ only in the leading underscore -- that seems like a setup for a pratfall.

Oh, and, let's check my `bash` fu again...how do you like this?
```
        for st in ${_agent_subtst_list}; do
            st_name=$(basename ${st})
            if [[ "${subtst:-${st_name}}" == "${st_name}"]]; then
                echo "${_major} ${st} ${posargs}"
            fi
        done > ${_para_jobs_file}
```
😁

File: exec-unittests
Comment (webbnh) [inline exec-unittests:192]: I recommend `-n ${_ECHO}` instead of comparing its value to `echo`.

And, I would also place this `if` statement before the block comment.  (We test `_para_jobs_file` on the previous line, and the comment refers to the line after this block, so let's move this block out of the way.)

File: exec-unittests
Comment (webbnh) [inline exec-unittests:209]: While this check for `python` doesn't hurt anything, as things stand at the moment, it doesn't really help anything either:  if `${subtst}` is `python`, then we won't enter either of the `if`'s at lines 211 or 217 and we'll just fall out the bottom anyway.  (And, if you accept my suggestion to get rid of the loop, then the `continue` won't be available to get us out of here.)

**Review by webbnh — DISMISSED** [review]
Looks good.

I got some suggestions and recommendations for you to consider.

**Review by dbutenhof — APPROVED** [review]

File: exec-unittests
Comment (portante) [reply exec-unittests:178]: I don't see the value in that, nor do I believe it is confusing.

Since this is causing this PR to be delayed, I'll just remove its use.

**Review by dbutenhof — APPROVED** [review]

**Review by webbnh — DISMISSED** [review]
Looks good...except for one thing....

However, if you feel like this falls below the line, just let me know and I'll resolve the comment.

**Review by webbnh — APPROVED** [review]
👍

**Review by npalaska — APPROVED** [review]

**Review by ndokos — APPROVED** [review]

**Pull request merged**
